platform-sdk/swirlds-merkledb/src/main/java/com/swirlds/merkledb/collections/LongListHeap.java

```diff
@@ -87,23 +91,34 @@ public LongListHeap(final Path file, final Configuration configuration) throws I
 
     /** {@inheritDoc} */
     @Override
-    protected void readBodyFromFileChannelOnInit(String sourceFileName, FileChannel fileChannel) throws IOException {
-        // read data
-        final int numOfArrays = calculateNumberOfChunks(size());
-        final ByteBuffer buffer = allocateDirect(memoryChunkSize);
-        buffer.order(ByteOrder.nativeOrder());
-        for (int i = 0; i < numOfArrays; i++) {
-            final AtomicLongArray atomicLongArray = new AtomicLongArray(numLongsPerChunk);
-            buffer.clear();
-            MerkleDbFileUtils.completelyRead(fileChannel, buffer);
-            buffer.flip();
-            int index = 0;
-            while (buffer.remaining() > 0) {
-                atomicLongArray.set(index, buffer.getLong());
-                index++;
-            }
-            chunkList.set(i, atomicLongArray);
+    protected void readBodyFromFileChannelOnInit(final String sourceFileName, final FileChannel fileChannel)
+            throws IOException {
+        initReadBuffer = ByteBuffer.allocateDirect(memoryChunkSize).order(ByteOrder.nativeOrder());
+
+        super.readBodyFromFileChannelOnInit(sourceFileName, fileChannel);
+
+        try {
+            MemoryUtils.closeDirectByteBuffer(initReadBuffer);
+        } finally {
+            initReadBuffer = null;
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    protected AtomicLongArray readChunkData(FileChannel fileChannel, int chunkIndex, int startIndex, int endIndex)
+            throws IOException {
+        AtomicLongArray chunk = createChunk();
+
+        readDataIntoBuffer(fileChannel, chunkIndex, startIndex, endIndex, initReadBuffer);
+        initReadBuffer.flip();
```

Comment:
`flip()` sets buffer position to 0. Should it be set to `startIndex * Long.BYTES` instead?

Reply:
> Should it be set to startIndex * Long.BYTES instead?

No, I just double checked it.

Reply:
OK, I see. We set longs in the chunk starting from index 0. If index started from startIndex below, position would be startIndex * Long.BYTES.

Reply:
Done: https://github.com/hashgraph/hedera-services/pull/16994/commits/5e66c44239883a093bfd4f6faefefe0adb513e12
cc @imalygin